### PR TITLE
Transaction notifications

### LINKED ISSUE #4203: Show browser notifications when transactions complete.
We should use the browser notification API to let the user know when a transaction enters a [final state]:

- confirmed
- failed
- dropped

Those three confirmations would have their own texts:

### Confirmed tx

Transaction ${tx.nonce} confirmed! `View on Etherscan`

### Failed tx

Transaction ${tx.nonce} failed. (Maybe append tx.error.message)

### Dropped tx

A Transaction ${tx.nonce} was dropped, because another transaction with that number was successfully processed.

(We may not want to show this one, especially since it is usually accompanied by a confirmed tx)

## Implementation

Should use [extensionizer](https://www.npmjs.com/package/extensionizer) to access the notifications API:

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/notifications

MetaMask does not include platform-specific APIs within the [metamask-controller](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/metamask-controller.js) or its dependencies. Instead, functions to `showConfirmedTransaction`, `showFailedTransaction`, and `showDroppedTransaction` should be defined on the [extension platform object](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/platforms/extension.js), which is passed in to the MetaMask controller on initialization:
https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L261

### PR DESCRIPTION
**Pull request by scsaba**
Fixes #4203

### CODE REVIEW

**Comment by scsaba** [comment]
Hi @danfinlay 
Could you please review this PR?

I think notification for dropped transactions is also informative, so I added that also. However it can cause problems in Firefox as mentioned [in the documentation](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/notifications/create):

> If you call notifications.create() more than once in rapid succession, Firefox may end up not displaying any notification at all.

I experienced this a few times.

Also, about "View on EtherScan" on the confirmed transaction's notification: it cannot be added as a link, so I had to handle the notifications.OnClicked event to open the transaction details on a new tab.

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -164,6 +164,13 @@ module.exports = class MetamaskController extends EventEmitter {
     })
     this.txController.on('newUnapprovedTx', opts.showUnapprovedTx.bind(opts))
 
+    this.txController.on(`tx:status-update`, (txId, status) => {
+      if (status === 'confirmed' || status === 'failed' || status === 'dropped') {
+        const txMeta = this.txController.txStateManager.getTx(txId)
+        this.platform.showTransactionNotification(txMeta)
```
Comment (danfinlay) [inline app/scripts/metamask-controller.js]: I like the generic handler.

File: app/scripts/platforms/extension.js [hunk 2]
```diff
@@ -31,6 +32,69 @@ class ExtensionPlatform {
...
+
+  showTransactionNotification (txMeta) {
+
+    const status = txMeta.status
+    if (status === 'confirmed') {
+      this._showConfirmedTransaction(txMeta)
+    } else if (status === 'failed') {
+      this._showFailedTransaction(txMeta)
+    } else if (status === 'dropped') {
+      this._showDroppedTransaction(txMeta)
```
Comment (danfinlay) [inline app/scripts/platforms/extension.js]: I was already on the fence about showing dropped txs, but if Firefox risks not showing any, I think I'd rather remove the dropped handler.

**Review by danfinlay — CHANGES_REQUESTED** [review]
Thanks so much, @scsaba!

Two small changes I would request:

- Let's remove the dropped tx handler, the firefox issue puts it over the top for now.
- Please add a changelog entry under the `master` header, noting `Now shows notifications when transactions are completed.`

**Comment by scsaba** [comment]
@danfinlay thank you for the quick review. I applied the requested changes. I was not sure about the changelog entry if I need to add the pull request nr also, I hope this is the correct version :)

**Review by danfinlay — APPROVED** [review]
Great job, thank you!

**Pull request merged**